{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "46ab290b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import necessary libraries for the analysis\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9565ae15",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Functions used below are imported from another notebook, run before changing directory\n",
    "\n",
    "%run overlap_and_range_functions.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c8d19589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1    chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2    chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3    chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4    chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode = pd.read_table(\"/Users/ryanyutian/Desktop/annotation_dataset/gencode.v43.annotation.gff3\", comment=\"#\",\n",
    "                        sep = \"\\t\", names = ['seqname', 'source', 'feature', 'start' , 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b10dcb28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',\n",
       "       'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21',\n",
       "       'chr22', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9',\n",
       "       'chrM', 'chrX', 'chrY'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['seqname'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "22b15ef4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CDS', 'exon', 'five_prime_UTR', 'gene', 'start_codon',\n",
       "       'stop_codon', 'stop_codon_redefined_as_selenocysteine',\n",
       "       'three_prime_UTR', 'transcript'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['feature'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9be335ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=ENSG00000223972.6;gene_id=ENSG00000223972.6;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode.iloc[5]['attribute']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8b088b25",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_info(x):\n",
    "    \n",
    "    if 'Parent' in x:\n",
    "        \n",
    "        parent = list(filter(lambda x: 'Parent' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        parent = 'N/A'\n",
    "        \n",
    "    g_id = list(filter(lambda x: 'gene_id' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_level = int(list(filter(lambda x: 'level' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    \n",
    "    return (parent, g_id, g_name, g_type, g_level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "383877e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1    chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2    chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3    chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4    chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute             parent  \\\n",
       "0  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...                N/A   \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  ENSG00000290825.1   \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  ENST00000456328.2   \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  ENST00000456328.2   \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  ENST00000456328.2   \n",
       "\n",
       "             gene_id gene_name gene_type  gene_level  \n",
       "0  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "1  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "2  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "3  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "4  ENSG00000290825.1   DDX11L2    lncRNA           2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode['parent'], gencode['gene_id'], gencode['gene_name'], gencode['gene_type'], gencode['gene_level'] = \\\n",
    "zip(*gencode.attribute.apply(lambda x: gene_info(x)))\n",
    "\n",
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "eda5c7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode = gencode.rename(columns={'seqname': 'chr'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e5362306",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['IG_C_gene', 'IG_C_pseudogene', 'IG_D_gene', 'IG_J_gene',\n",
       "       'IG_J_pseudogene', 'IG_V_gene', 'IG_V_pseudogene', 'IG_pseudogene',\n",
       "       'Mt_rRNA', 'Mt_tRNA', 'TEC', 'TR_C_gene', 'TR_D_gene', 'TR_J_gene',\n",
       "       'TR_J_pseudogene', 'TR_V_gene', 'TR_V_pseudogene', 'artifact',\n",
       "       'lncRNA', 'miRNA', 'misc_RNA', 'processed_pseudogene',\n",
       "       'protein_coding', 'pseudogene', 'rRNA', 'rRNA_pseudogene',\n",
       "       'ribozyme', 'sRNA', 'scRNA', 'scaRNA', 'snRNA', 'snoRNA',\n",
       "       'transcribed_processed_pseudogene',\n",
       "       'transcribed_unitary_pseudogene',\n",
       "       'transcribed_unprocessed_pseudogene',\n",
       "       'translated_processed_pseudogene',\n",
       "       'translated_unprocessed_pseudogene', 'unitary_pseudogene',\n",
       "       'unprocessed_pseudogene', 'vault_RNA'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['gene_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "538e9424",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CDS', 'exon', 'five_prime_UTR', 'gene', 'start_codon',\n",
       "       'stop_codon', 'stop_codon_redefined_as_selenocysteine',\n",
       "       'three_prime_UTR', 'transcript'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(gencode['feature'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "81d863c5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source     feature  start    end score strand frame  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute             parent  \\\n",
       "0  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...                N/A   \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  ENSG00000290825.1   \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  ENST00000456328.2   \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  ENST00000456328.2   \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  ENST00000456328.2   \n",
       "\n",
       "             gene_id gene_name gene_type  gene_level  \n",
       "0  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "1  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "2  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "3  ENSG00000290825.1   DDX11L2    lncRNA           2  \n",
       "4  ENSG00000290825.1   DDX11L2    lncRNA           2  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "776fb76c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>percent_overlap</th>\n",
       "      <th>input_start</th>\n",
       "      <th>input_end</th>\n",
       "      <th>input_ind</th>\n",
       "      <th>input_sv_type</th>\n",
       "      <th>input_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2558</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206592</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000186891.14;gene_id=ENSG00000186891....</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000186891.14</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2559</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000328596.10;Parent=ENSG00000186891.1...</td>\n",
       "      <td>ENSG00000186891.14</td>\n",
       "      <td>ENSG00000186891.14</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2560</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1206385</td>\n",
       "      <td>1206571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000328596.10:1;Parent=ENST000003...</td>\n",
       "      <td>ENST00000328596.10</td>\n",
       "      <td>ENSG00000186891.14</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1206385</td>\n",
       "      <td>1206571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000328596.10;Parent=ENST000003285...</td>\n",
       "      <td>ENST00000328596.10</td>\n",
       "      <td>ENSG00000186891.14</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2562</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1206569</td>\n",
       "      <td>1206571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=start_codon:ENST00000328596.10;Parent=ENST0...</td>\n",
       "      <td>ENST00000328596.10</td>\n",
       "      <td>ENSG00000186891.14</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3609</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1299973</td>\n",
       "      <td>1300068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000472541.5:8;Parent=ENST0000047...</td>\n",
       "      <td>ENST00000472541.5</td>\n",
       "      <td>ENSG00000131584.19</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>29.17</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3610</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1298744</td>\n",
       "      <td>1299905</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000472541.5:9;Parent=ENST0000047...</td>\n",
       "      <td>ENST00000472541.5</td>\n",
       "      <td>ENSG00000131584.19</td>\n",
       "      <td>ACAP3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3647</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>1296110</td>\n",
       "      <td>1296170</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000278073.1;gene_id=ENSG00000278073.1...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000278073.1</td>\n",
       "      <td>MIR6726</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3648</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1296110</td>\n",
       "      <td>1296170</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000613751.1;Parent=ENSG00000278073.1;...</td>\n",
       "      <td>ENSG00000278073.1</td>\n",
       "      <td>ENSG00000278073.1</td>\n",
       "      <td>MIR6726</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3649</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1296110</td>\n",
       "      <td>1296170</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000613751.1:1;Parent=ENST0000061...</td>\n",
       "      <td>ENST00000613751.1</td>\n",
       "      <td>ENSG00000278073.1</td>\n",
       "      <td>MIR6726</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1200000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1</td>\n",
       "      <td>del</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1016 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr   source      feature    start      end score strand frame  \\\n",
       "2558  chr1   HAVANA         gene  1203508  1206592     .      -     .   \n",
       "2559  chr1   HAVANA   transcript  1203508  1206571     .      -     .   \n",
       "2560  chr1   HAVANA         exon  1206385  1206571     .      -     .   \n",
       "2561  chr1   HAVANA          CDS  1206385  1206571     .      -     0   \n",
       "2562  chr1   HAVANA  start_codon  1206569  1206571     .      -     0   \n",
       "...    ...      ...          ...      ...      ...   ...    ...   ...   \n",
       "3609  chr1   HAVANA         exon  1299973  1300068     .      -     .   \n",
       "3610  chr1   HAVANA         exon  1298744  1299905     .      -     .   \n",
       "3647  chr1  ENSEMBL         gene  1296110  1296170     .      -     .   \n",
       "3648  chr1  ENSEMBL   transcript  1296110  1296170     .      -     .   \n",
       "3649  chr1  ENSEMBL         exon  1296110  1296170     .      -     .   \n",
       "\n",
       "                                              attribute              parent  \\\n",
       "2558  ID=ENSG00000186891.14;gene_id=ENSG00000186891....                 N/A   \n",
       "2559  ID=ENST00000328596.10;Parent=ENSG00000186891.1...  ENSG00000186891.14   \n",
       "2560  ID=exon:ENST00000328596.10:1;Parent=ENST000003...  ENST00000328596.10   \n",
       "2561  ID=CDS:ENST00000328596.10;Parent=ENST000003285...  ENST00000328596.10   \n",
       "2562  ID=start_codon:ENST00000328596.10;Parent=ENST0...  ENST00000328596.10   \n",
       "...                                                 ...                 ...   \n",
       "3609  ID=exon:ENST00000472541.5:8;Parent=ENST0000047...   ENST00000472541.5   \n",
       "3610  ID=exon:ENST00000472541.5:9;Parent=ENST0000047...   ENST00000472541.5   \n",
       "3647  ID=ENSG00000278073.1;gene_id=ENSG00000278073.1...                 N/A   \n",
       "3648  ID=ENST00000613751.1;Parent=ENSG00000278073.1;...   ENSG00000278073.1   \n",
       "3649  ID=exon:ENST00000613751.1:1;Parent=ENST0000061...   ENST00000613751.1   \n",
       "\n",
       "                 gene_id gene_name       gene_type  gene_level  \\\n",
       "2558  ENSG00000186891.14  TNFRSF18  protein_coding           2   \n",
       "2559  ENSG00000186891.14  TNFRSF18  protein_coding           2   \n",
       "2560  ENSG00000186891.14  TNFRSF18  protein_coding           2   \n",
       "2561  ENSG00000186891.14  TNFRSF18  protein_coding           2   \n",
       "2562  ENSG00000186891.14  TNFRSF18  protein_coding           2   \n",
       "...                  ...       ...             ...         ...   \n",
       "3609  ENSG00000131584.19     ACAP3  protein_coding           2   \n",
       "3610  ENSG00000131584.19     ACAP3  protein_coding           2   \n",
       "3647   ENSG00000278073.1   MIR6726           miRNA           3   \n",
       "3648   ENSG00000278073.1   MIR6726           miRNA           3   \n",
       "3649   ENSG00000278073.1   MIR6726           miRNA           3   \n",
       "\n",
       "      percent_overlap  input_start  input_end  input_ind input_sv_type  \\\n",
       "2558           100.00      1200000    1300000          1           del   \n",
       "2559           100.00      1200000    1300000          1           del   \n",
       "2560           100.00      1200000    1300000          1           del   \n",
       "2561           100.00      1200000    1300000          1           del   \n",
       "2562           100.00      1200000    1300000          1           del   \n",
       "...               ...          ...        ...        ...           ...   \n",
       "3609            29.17      1200000    1300000          1           del   \n",
       "3610           100.00      1200000    1300000          1           del   \n",
       "3647           100.00      1200000    1300000          1           del   \n",
       "3648           100.00      1200000    1300000          1           del   \n",
       "3649           100.00      1200000    1300000          1           del   \n",
       "\n",
       "     input_sample_id  \n",
       "2558          sample  \n",
       "2559          sample  \n",
       "2560          sample  \n",
       "2561          sample  \n",
       "2562          sample  \n",
       "...              ...  \n",
       "3609          sample  \n",
       "3610          sample  \n",
       "3647          sample  \n",
       "3648          sample  \n",
       "3649          sample  \n",
       "\n",
       "[1016 rows x 20 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_func('chr1', 1200000, 1300000, 1, 'del', 'sample', gencode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48742a7c",
   "metadata": {},
   "source": [
    "# Small dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a768d5fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load filtered somatic small deletions ###\n",
    "\n",
    "##\n",
    "somatic_small_del_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/small_dels'\n",
    "\n",
    "somatic_small_del_wo_highoverlap_df_names = []\n",
    "\n",
    "os.chdir(somatic_small_del_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_small_del_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_small_del_wo_highoverlap_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7e9ce6bd",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently analyzing: A_RR_GBM809\n",
      "Currently analyzing: A_R_GBM607\n",
      "Currently analyzing: B_P_GBM593\n",
      "Currently analyzing: B_R_GBM898\n",
      "Currently analyzing: C_P_GBM577\n",
      "Currently analyzing: C_R_GBM625\n",
      "Currently analyzing: E_RR_GBM937\n",
      "Currently analyzing: E_R_GBM781\n",
      "Currently analyzing: F_P_GBM620\n",
      "Currently analyzing: F_R_GBM691\n",
      "Currently analyzing: G_P_GBM454\n",
      "Currently analyzing: G_R_GBM833\n",
      "Currently analyzing: H_P_GBM460\n",
      "Currently analyzing: H_R_GBM492\n",
      "Currently analyzing: I_P_GBM440\n",
      "Currently analyzing: I_R_GBM532\n",
      "Currently analyzing: J_P_GBM401\n",
      "Currently analyzing: J_RR_GBM551\n",
      "Currently analyzing: J_R_GBM498\n",
      "Currently analyzing: K_P_GBM529\n",
      "Currently analyzing: K_R_GBM832\n",
      "Currently analyzing: L_P_GBM618\n",
      "Currently analyzing: L_R_SMTB152\n",
      "Currently analyzing: M_P_GBM672\n",
      "Currently analyzing: M_R_GBM828\n",
      "Currently analyzing: N_P_BT2013110\n",
      "Currently analyzing: N_R_GBM745\n",
      "Currently analyzing: O_P_GBM703\n",
      "Currently analyzing: O_R_SMTB781\n",
      "Currently analyzing: P_P_SMTB123\n",
      "Currently analyzing: P_R_SMTB265\n",
      "Currently analyzing: Q_P_SMTB665\n",
      "Currently analyzing: Q_R_GBM1070\n",
      "Currently analyzing: X_R_GBM945\n",
      "Currently analyzing: X_R_SMTB135\n",
      "Currently analyzing: X_R_SMTB241\n",
      "Currently analyzing: X_R_SMTB302\n",
      "Currently analyzing: X_R_SMTB814\n"
     ]
    }
   ],
   "source": [
    "small_del_gene_list_names = []\n",
    "\n",
    "for small_del_df_name in somatic_small_del_wo_highoverlap_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + small_del_df_name[:-41])\n",
    "    temp_df = globals()[small_del_df_name]\n",
    "    temp_output_df = pd.DataFrame()\n",
    "    \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_overlap = overlap_func(row['CHROM'], row['POS'], row['END'], index, \\\n",
    "                                    'del', small_del_df_name[:-41], gencode)\n",
    "        temp_output_df = temp_output_df.append(temp_overlap)\n",
    "    \n",
    "    temp_gene_list_name = small_del_df_name[:-41] + '_small_del_gene_list'\n",
    "    small_del_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fc1ff591",
   "metadata": {},
   "outputs": [],
   "source": [
    "small_del_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/small_dels/'\n",
    "\n",
    "for df_name in small_del_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((small_del_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19c35fae",
   "metadata": {},
   "source": [
    "# Large dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5299d5b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## DELs\n",
    "somatic_large_DEL_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/large_svs/DEL'\n",
    "\n",
    "somatic_large_DEL_wo_highoverlap_df_names = []\n",
    "\n",
    "os.chdir(somatic_large_DEL_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_large_DEL_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_large_DEL_wo_highoverlap_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9927aa32",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently analyzing: A_RR_GBM809\n",
      "Currently analyzing: A_R_GBM607\n",
      "Currently analyzing: B_P_GBM593\n",
      "Currently analyzing: B_R_GBM898\n",
      "Currently analyzing: C_P_GBM577\n",
      "Currently analyzing: C_R_GBM625\n",
      "Currently analyzing: E_RR_GBM937\n",
      "Currently analyzing: E_R_GBM781\n",
      "Currently analyzing: F_P_GBM620\n",
      "Currently analyzing: F_R_GBM691\n",
      "Currently analyzing: G_P_GBM454\n",
      "Currently analyzing: G_R_GBM833\n",
      "Currently analyzing: H_P_GBM460\n",
      "Currently analyzing: H_R_GBM492\n",
      "Currently analyzing: I_P_GBM440\n",
      "Currently analyzing: I_R_GBM532\n",
      "Currently analyzing: J_P_GBM401\n",
      "Currently analyzing: J_RR_GBM551\n",
      "Currently analyzing: J_R_GBM498\n",
      "Currently analyzing: K_P_GBM529\n",
      "Currently analyzing: K_R_GBM832\n",
      "Currently analyzing: L_P_GBM618\n",
      "Currently analyzing: L_R_SMTB152\n",
      "Currently analyzing: M_P_GBM672\n",
      "Currently analyzing: M_R_GBM828\n",
      "Currently analyzing: N_P_BT2013110\n",
      "Currently analyzing: N_R_GBM745\n",
      "Currently analyzing: O_P_GBM703\n",
      "Currently analyzing: O_R_SMTB781\n",
      "Currently analyzing: P_P_SMTB123\n",
      "Currently analyzing: P_R_SMTB265\n",
      "Currently analyzing: Q_P_SMTB665\n",
      "Currently analyzing: Q_R_GBM1070\n",
      "Currently analyzing: X_R_GBM945\n",
      "Currently analyzing: X_R_SMTB135\n",
      "Currently analyzing: X_R_SMTB241\n",
      "Currently analyzing: X_R_SMTB302\n",
      "Currently analyzing: X_R_SMTB814\n"
     ]
    }
   ],
   "source": [
    "large_del_gene_list_names = []\n",
    "\n",
    "for large_del_df_name in somatic_large_DEL_wo_highoverlap_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + large_del_df_name[:-44])\n",
    "    temp_df = globals()[large_del_df_name]\n",
    "    temp_output_df = pd.DataFrame()\n",
    "    \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_overlap = overlap_func(row['CHROM'], row['POS'], row['END'], index, \\\n",
    "                                    'DEL', large_del_df_name[:-44], gencode)\n",
    "        temp_output_df = temp_output_df.append(temp_overlap)\n",
    "    \n",
    "    temp_gene_list_name = large_del_df_name[:-44] + '_large_del_gene_list'\n",
    "    large_del_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "64c08e3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_del_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/large_svs/DEL/'\n",
    "\n",
    "for df_name in large_del_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((large_del_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf8f13f0",
   "metadata": {},
   "source": [
    "# Large dups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "aba90ec3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## DUPs\n",
    "somatic_large_DUP_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/large_svs/DUP'\n",
    "\n",
    "somatic_large_DUP_wo_highoverlap_df_names = []\n",
    "\n",
    "os.chdir(somatic_large_DUP_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_large_DUP_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_large_DUP_wo_highoverlap_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "74d640ea",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently analyzing: A_RR_GBM809\n",
      "Currently analyzing: A_R_GBM607\n",
      "Currently analyzing: B_P_GBM593\n",
      "Currently analyzing: B_R_GBM898\n",
      "Currently analyzing: C_P_GBM577\n",
      "Currently analyzing: C_R_GBM625\n",
      "Currently analyzing: E_RR_GBM937\n",
      "Currently analyzing: E_R_GBM781\n",
      "Currently analyzing: F_P_GBM620\n",
      "Currently analyzing: F_R_GBM691\n",
      "Currently analyzing: G_P_GBM454\n",
      "Currently analyzing: G_R_GBM833\n",
      "Currently analyzing: H_P_GBM460\n",
      "Currently analyzing: H_R_GBM492\n",
      "Currently analyzing: I_P_GBM440\n",
      "Currently analyzing: I_R_GBM532\n",
      "Currently analyzing: J_P_GBM401\n",
      "Currently analyzing: J_RR_GBM551\n",
      "Currently analyzing: J_R_GBM498\n",
      "Currently analyzing: K_P_GBM529\n",
      "Currently analyzing: K_R_GBM832\n",
      "Currently analyzing: L_P_GBM618\n",
      "Currently analyzing: L_R_SMTB152\n",
      "Currently analyzing: M_P_GBM672\n",
      "Currently analyzing: M_R_GBM828\n",
      "Currently analyzing: N_P_BT2013110\n",
      "Currently analyzing: N_R_GBM745\n",
      "Currently analyzing: O_P_GBM703\n",
      "Currently analyzing: O_R_SMTB781\n",
      "Currently analyzing: P_P_SMTB123\n",
      "Currently analyzing: P_R_SMTB265\n",
      "Currently analyzing: Q_P_SMTB665\n",
      "Currently analyzing: Q_R_GBM1070\n",
      "Currently analyzing: X_R_GBM945\n",
      "Currently analyzing: X_R_SMTB135\n",
      "Currently analyzing: X_R_SMTB241\n",
      "Currently analyzing: X_R_SMTB302\n",
      "Currently analyzing: X_R_SMTB814\n"
     ]
    }
   ],
   "source": [
    "large_dup_gene_list_names = []\n",
    "\n",
    "for large_dup_df_name in somatic_large_DUP_wo_highoverlap_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + large_dup_df_name[:-44])\n",
    "    temp_df = globals()[large_dup_df_name]\n",
    "    temp_output_df = pd.DataFrame()\n",
    "   \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_overlap = overlap_func(row['CHROM'], row['POS'], row['END'], index, \\\n",
    "                                    'DUP', large_dup_df_name[:-44], gencode)\n",
    "        temp_output_df = temp_output_df.append(temp_overlap)\n",
    "    \n",
    "    temp_gene_list_name = large_dup_df_name[:-44] + '_large_dup_gene_list'\n",
    "    large_dup_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4b025f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_dup_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/large_svs/DUP/'\n",
    "\n",
    "for df_name in large_dup_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((large_dup_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589bb70d",
   "metadata": {},
   "source": [
    "# Large invs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "59cc9e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## INVs\n",
    "somatic_large_INV_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_glioma_sv_processed/somatic_panel_final/large_svs/INV'\n",
    "\n",
    "somatic_large_INV_wo_highoverlap_df_names = []\n",
    "\n",
    "os.chdir(somatic_large_INV_path)\n",
    "temp_files = sorted([i for i in os.listdir(somatic_large_INV_path) if 'DS' not in i])\n",
    "\n",
    "for file_name in temp_files:\n",
    "    \n",
    "    globals()[file_name[:-4]] = pd.read_csv(file_name)\n",
    "    somatic_large_INV_wo_highoverlap_df_names.append(file_name[:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1537935c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently analyzing: A_RR_GBM809\n",
      "Currently analyzing: A_R_GBM607\n",
      "Currently analyzing: B_P_GBM593\n",
      "Currently analyzing: B_R_GBM898\n",
      "Currently analyzing: C_P_GBM577\n",
      "Currently analyzing: C_R_GBM625\n",
      "Currently analyzing: E_RR_GBM937\n",
      "Currently analyzing: E_R_GBM781\n",
      "Currently analyzing: F_P_GBM620\n",
      "Currently analyzing: F_R_GBM691\n",
      "Currently analyzing: G_P_GBM454\n",
      "Currently analyzing: G_R_GBM833\n",
      "Currently analyzing: H_P_GBM460\n",
      "Currently analyzing: H_R_GBM492\n",
      "Currently analyzing: I_P_GBM440\n",
      "Currently analyzing: I_R_GBM532\n",
      "Currently analyzing: J_P_GBM401\n",
      "Currently analyzing: J_RR_GBM551\n",
      "Currently analyzing: J_R_GBM498\n",
      "Currently analyzing: K_P_GBM529\n",
      "Currently analyzing: K_R_GBM832\n",
      "Currently analyzing: L_P_GBM618\n",
      "Currently analyzing: L_R_SMTB152\n",
      "Currently analyzing: M_P_GBM672\n",
      "Currently analyzing: M_R_GBM828\n",
      "Currently analyzing: N_P_BT2013110\n",
      "Currently analyzing: N_R_GBM745\n",
      "Currently analyzing: O_P_GBM703\n",
      "Currently analyzing: O_R_SMTB781\n",
      "Currently analyzing: P_P_SMTB123\n",
      "Currently analyzing: P_R_SMTB265\n",
      "Currently analyzing: Q_P_SMTB665\n",
      "Currently analyzing: Q_R_GBM1070\n",
      "Currently analyzing: X_R_GBM945\n",
      "Currently analyzing: X_R_SMTB135\n",
      "Currently analyzing: X_R_SMTB241\n",
      "Currently analyzing: X_R_SMTB302\n",
      "Currently analyzing: X_R_SMTB814\n"
     ]
    }
   ],
   "source": [
    "large_inv_gene_list_names = []\n",
    "\n",
    "for large_inv_df_name in somatic_large_INV_wo_highoverlap_df_names:\n",
    "    \n",
    "    print('Currently analyzing: ' + large_inv_df_name[:-44])\n",
    "    temp_df = globals()[large_inv_df_name]\n",
    "    temp_output_df = pd.DataFrame()\n",
    "    \n",
    "    for index, row in temp_df.iterrows():\n",
    "        \n",
    "        temp_overlap = overlap_func(row['CHROM'], row['POS'], row['END'], index, \\\n",
    "                                    'INV', large_inv_df_name[:-44], gencode)\n",
    "        temp_output_df = temp_output_df.append(temp_overlap)\n",
    "    \n",
    "    temp_gene_list_name = large_inv_df_name[:-44] + '_large_inv_gene_list'\n",
    "    large_inv_gene_list_names.append(temp_gene_list_name)\n",
    "    \n",
    "    globals()[temp_gene_list_name] = temp_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "caa76cb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_inv_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/all/large_svs/INV/'\n",
    "\n",
    "for df_name in large_inv_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df.to_csv((large_inv_gene_list_path + df_name + '.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3c53e17",
   "metadata": {},
   "source": [
    "# COMBINED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a564f8d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "combined_gene_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/unique_gene_only_annotated/combined/'\n",
    "\n",
    "for df_name in small_del_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name].copy()\n",
    "    temp_df = temp_df.append(globals()[df_name[:-20]+'_large_del_gene_list'], ignore_index=True)\n",
    "    temp_df = temp_df.append(globals()[df_name[:-20]+'_large_dup_gene_list'], ignore_index=True)\n",
    "    temp_df = temp_df.append(globals()[df_name[:-20]+'_large_inv_gene_list'], ignore_index=True)\n",
    "\n",
    "    temp_df.to_csv((combined_gene_list_path + df_name[:-20] + '.csv'), index=False, sep=',')\n",
    "    \n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d831742e",
   "metadata": {},
   "source": [
    "# GENE ONLY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "144fd85b",
   "metadata": {},
   "outputs": [],
   "source": [
    "small_del_gene_only_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/gene_only/small_dels/'\n",
    "\n",
    "for df_name in small_del_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    temp_df_gene_only = temp_df[temp_df['feature'] == 'gene']\n",
    "    \n",
    "    temp_df_gene_only.to_csv((small_del_gene_only_list_path + df_name[:-20] + '_small_del_gene_list.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6ab4a627",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_del_gene_only_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/gene_only/large_svs/DEL/'\n",
    "\n",
    "for df_name in large_del_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "\n",
    "    if len(temp_df) != 0:\n",
    "        \n",
    "        temp_df_gene_only = temp_df[temp_df['feature'] == 'gene']\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        temp_df_gene_only = temp_df\n",
    "\n",
    "    \n",
    "    temp_df_gene_only.to_csv((large_del_gene_only_list_path + df_name[:-20] + '_large_del_gene_list.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b69790da",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_dup_gene_only_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/gene_only/large_svs/DUP/'\n",
    "\n",
    "for df_name in large_dup_gene_list_names:\n",
    "\n",
    "    temp_df = globals()[df_name]\n",
    "\n",
    "    if len(temp_df) != 0:\n",
    "        \n",
    "        temp_df_gene_only = temp_df[temp_df['feature'] == 'gene']\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        temp_df_gene_only = temp_df\n",
    "\n",
    "    \n",
    "    temp_df_gene_only.to_csv((large_dup_gene_only_list_path + df_name[:-20] + '_large_dup_gene_list.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "283e2485",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_inv_gene_only_list_path = '/Users/ryanyutian/Desktop/new_panel_TRI_Brain_gene_list/gene_only/large_svs/INV/'\n",
    "\n",
    "for df_name in large_inv_gene_list_names:\n",
    "    \n",
    "    temp_df = globals()[df_name]\n",
    "    \n",
    "    if len(temp_df) != 0:\n",
    "        \n",
    "        temp_df_gene_only = temp_df[temp_df['feature'] == 'gene']\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        temp_df_gene_only = temp_df\n",
    "        \n",
    "    temp_df_gene_only.to_csv((large_inv_gene_only_list_path + df_name[:-20] + '_large_inv_gene_list.csv'), index=False, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad5642f5",
   "metadata": {},
   "source": [
    "# INSPECTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cd10617b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>508754</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>211375717</td>\n",
       "      <td>212538841</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000178568.15;gene_id=ENSG00000178568....</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508755</th>\n",
       "      <td>chr2</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>211375717</td>\n",
       "      <td>212124907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000402597.6;Parent=ENSG00000178568.15...</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508756</th>\n",
       "      <td>chr2</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>212124752</td>\n",
       "      <td>212124907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000402597.6:1;Parent=ENST0000040...</td>\n",
       "      <td>ENST00000402597.6</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508757</th>\n",
       "      <td>chr2</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>212124752</td>\n",
       "      <td>212124907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000402597.6;Parent=ENST0000040259...</td>\n",
       "      <td>ENST00000402597.6</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508758</th>\n",
       "      <td>chr2</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>211947430</td>\n",
       "      <td>211947616</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000402597.6:2;Parent=ENST0000040...</td>\n",
       "      <td>ENST00000402597.6</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509011</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>212429159</td>\n",
       "      <td>212429247</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR5:ENST00000435846.1;Parent=ENST000004358...</td>\n",
       "      <td>ENST00000435846.1</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509012</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>212124809</td>\n",
       "      <td>212124903</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR5:ENST00000435846.1;Parent=ENST000004358...</td>\n",
       "      <td>ENST00000435846.1</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509013</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>212124584</td>\n",
       "      <td>212125181</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000459774.1;Parent=ENSG00000178568.15...</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509014</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>212125058</td>\n",
       "      <td>212125181</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000459774.1:1;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000459774.1</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509015</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>212124584</td>\n",
       "      <td>212124903</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000459774.1:2;Parent=ENST0000045...</td>\n",
       "      <td>ENST00000459774.1</td>\n",
       "      <td>ENSG00000178568.15</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>262 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr   source         feature      start        end score strand  \\\n",
       "508754  chr2   HAVANA            gene  211375717  212538841     .      -   \n",
       "508755  chr2  ENSEMBL      transcript  211375717  212124907     .      -   \n",
       "508756  chr2  ENSEMBL            exon  212124752  212124907     .      -   \n",
       "508757  chr2  ENSEMBL             CDS  212124752  212124907     .      -   \n",
       "508758  chr2  ENSEMBL            exon  211947430  211947616     .      -   \n",
       "...      ...      ...             ...        ...        ...   ...    ...   \n",
       "509011  chr2   HAVANA  five_prime_UTR  212429159  212429247     .      -   \n",
       "509012  chr2   HAVANA  five_prime_UTR  212124809  212124903     .      -   \n",
       "509013  chr2   HAVANA      transcript  212124584  212125181     .      -   \n",
       "509014  chr2   HAVANA            exon  212125058  212125181     .      -   \n",
       "509015  chr2   HAVANA            exon  212124584  212124903     .      -   \n",
       "\n",
       "       frame                                          attribute  \\\n",
       "508754     .  ID=ENSG00000178568.15;gene_id=ENSG00000178568....   \n",
       "508755     .  ID=ENST00000402597.6;Parent=ENSG00000178568.15...   \n",
       "508756     .  ID=exon:ENST00000402597.6:1;Parent=ENST0000040...   \n",
       "508757     0  ID=CDS:ENST00000402597.6;Parent=ENST0000040259...   \n",
       "508758     .  ID=exon:ENST00000402597.6:2;Parent=ENST0000040...   \n",
       "...      ...                                                ...   \n",
       "509011     .  ID=UTR5:ENST00000435846.1;Parent=ENST000004358...   \n",
       "509012     .  ID=UTR5:ENST00000435846.1;Parent=ENST000004358...   \n",
       "509013     .  ID=ENST00000459774.1;Parent=ENSG00000178568.15...   \n",
       "509014     .  ID=exon:ENST00000459774.1:1;Parent=ENST0000045...   \n",
       "509015     .  ID=exon:ENST00000459774.1:2;Parent=ENST0000045...   \n",
       "\n",
       "                    parent             gene_id gene_name       gene_type  \\\n",
       "508754                 N/A  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "508755  ENSG00000178568.15  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "508756   ENST00000402597.6  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "508757   ENST00000402597.6  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "508758   ENST00000402597.6  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "...                    ...                 ...       ...             ...   \n",
       "509011   ENST00000435846.1  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "509012   ENST00000435846.1  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "509013  ENSG00000178568.15  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "509014   ENST00000459774.1  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "509015   ENST00000459774.1  ENSG00000178568.15     ERBB4  protein_coding   \n",
       "\n",
       "        gene_level  \n",
       "508754           2  \n",
       "508755           3  \n",
       "508756           3  \n",
       "508757           3  \n",
       "508758           3  \n",
       "...            ...  \n",
       "509011           2  \n",
       "509012           2  \n",
       "509013           2  \n",
       "509014           2  \n",
       "509015           2  \n",
       "\n",
       "[262 rows x 14 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode[gencode['gene_name'].str.contains('ERBB4')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5f02cb01",
   "metadata": {},
   "outputs": [],
   "source": [
    "A_R_GBM607_df = A_R_GBM607_small_del_gene_list.copy()\n",
    "A_R_GBM607_df = A_R_GBM607_df.append(A_R_GBM607_large_del_gene_list, ignore_index=True)\n",
    "A_R_GBM607_df = A_R_GBM607_df.append(A_R_GBM607_large_dup_gene_list, ignore_index=True)\n",
    "A_R_GBM607_df = A_R_GBM607_df.append(A_R_GBM607_large_inv_gene_list, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ac96b1bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "A_R_GBM607_df_gene_only = A_R_GBM607_df[A_R_GBM607_df['feature'] == 'gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "eed54d8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>percent_overlap</th>\n",
       "      <th>input_start</th>\n",
       "      <th>input_end</th>\n",
       "      <th>input_ind</th>\n",
       "      <th>input_sv_type</th>\n",
       "      <th>input_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chr, source, feature, start, end, score, strand, frame, attribute, parent, gene_id, gene_name, gene_type, gene_level, percent_overlap, input_start, input_end, input_ind, input_sv_type, input_sample_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A_R_GBM607_df[A_R_GBM607_df['gene_name'].str.contains('ERBB4')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1215617d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
